Alerting/metrics

owen-d opened the pull request:
This PR adds a bunch of new metrics to the ngalert package and refactors it a bit. It reuses alertmanager libraries where possible.

Additionally, it fixes a mutex deadlock condition in the state iteration code and trims the active alert history to 100 _total_ on each iteration rather than removing the first 100. It also  moves the cleanup interval to 15s (prometheus scrape config) and includes metric reporting there.

owen-d committed:
moves alerting metrics to their own pkg

owen-d committed:
adds grafana_alerting_alerts (by state) metric

owen-d committed:
alerts_received_{total,invalid}

owen-d committed:
embed alertmanager alerting struct in ng metrics & remove duplicated notification metrics (already embed alertmanager notifier metrics)

owen-d committed:
use silence metrics from alertmanager lib

owen-d committed:
fix - manager has metrics

owen-d committed:
updates ngalert tests

owen-d committed:
comment lint
Signed-off-by: Owen Diehl <ow.diehl@gmail.com>

owen-d committed:
cleaner prom registry code

owen-d committed:
removes ngalert global metrics

owen-d committed:
new registry use in all tests

aknuds1 commented on pkg/tests/testinfra/testinfra.go:
```diff
@@ -23,6 +24,9 @@ import (
 // StartGrafana starts a Grafana server.
 // The server address is returned.
 func StartGrafana(t *testing.T, grafDir, cfgPath string, sqlStore *sqlstore.SQLStore) string {
+	// Grafana uses some global metric registration. Clear the
+	// registry between tests.
+	prometheus.DefaultRegisterer = prometheus.NewRegistry()
 	t.Helper()
```
Nit, but better to keep `t.Helper` at the very beginning of the function, because it matters when the stacktrace is calculated.
```suggestion
	t.Helper()
	// Grafana uses some global metric registration. Clear the
	// registry between tests.
	prometheus.DefaultRegisterer = prometheus.NewRegistry()
```

owen-d replied:
Gotcha,  thanks!

aknuds1 commented on pkg/services/ngalert/tests/manager_test.go:
```diff
@@ -16,6 +17,8 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
+var NilMetrics = metrics.NewMetrics(nil)
```
This shouldn't be exported should it?
```suggestion
var nilMetrics = metrics.NewMetrics(nil)
```

aknuds1 reviewed (approved):
LGTM, although I don't know the alerting code much. See nits though. There's also a problem with failing tests.

owen-d committed:
ngalert metrics impl service, hack testinfra code to prevent duplicate metric registrations

owen-d committed:
nilmetrics unexported

owen-d commented:
Thanks for the review. I've updated the PR and refactored it a bit by making the Metrics a service for dependency injection. There are a few hack-ish things due to metric registration panics in tests (the DI framework reuses the same prometheus registries across subsequent tests, which then re-register metrics). It wasn't feasible to make all the metrics global (nor is that a practice I love) as some of the libraries we're using from Alertmanager are creating their metrics in a non-global fashion.

gotjosh reviewed (approved):
LGTM
